File: scaler/RegExMatch.cpp
int Solution::isMatch(const string A, const string B) {
    int N = A.length(), M = B.length();
    bool DP[N + 1][M + 1];
    memset(DP, false, sizeof(DP));
    DP[0][0] = true;
    // Only '*' can match with empty string 
    for(int i = 1; i <= M; i++) {
        if(B[i - 1] == '*') {
            DP[0][i] = DP[0][i - 1];
        }
    }
    // fill the table
    for(int i = 1; i <= N; i++) {
        for(int j = 1; j <= M; j++) {
            if(B[j - 1] == '*') {
                DP[i][j] = DP[i][j - 1] || DP[i - 1][j];
            } else if(A[i - 1] == B[j - 1] || B[j - 1] == '?') {
                DP[i][j] = DP[i - 1][j - 1];
            } else {
                DP[i][j] = false;
            }
        }
    }
    return DP[N][M];
}

File: scaler/RegExMatch.java
public class Solution {
    
    Map<String, Integer> DP = new HashMap<>();
    
    int solve(String A, String B, int a, int b) {
        
        // base case
        if(a < 0 && b < 0) {
            return 1;
        }
        
        if(b < 0) {
            // not possible
            return 0;
        }
        
        char y = B.charAt(b);
        if(a < 0) {
            if(y == '*') {
                // try to solve recursively
                return solve(A, B, a, b - 1);
            } else {
                return 0; // not possible
            }
        }
        char x = A.charAt(a);
        
        // look into DP
        String state = a+"-"+b;
        if(DP.containsKey(state)) {
            // already computed
            return DP.get(state);
        }
        
        int flag = 0;
        if(x == y || y == '?') {
            return solve(A, B, a - 1, b - 1); // recursive call
        }
        if(y == '*') {
            // possible recursive call (because * means any sequence inlcluding empty sequence)
            for(int i = 0; i <= a + 1; i++) {
                if(solve(A, B, a - i, b - 1) == 1) {
                    // possible
                    flag = 1;
                    break;
                }
            }
        }
        // store it for further use
        DP.put(state, flag);
        
        return flag;
    }
    
    // DO NOT MODIFY THE ARGUMENTS WITH "final" PREFIX. IT IS READ ONLY
    public int isMatch(final String A, final String B) {
        return solve(A, B, A.length() - 1, B.length() - 1);
    }
}